src/api.js

```diff
@@ -844,6 +912,48 @@ Module["onRuntimeInitialized"] = function onRuntimeInitialized() {
         return stmt;
     };
 
+    /** Iterate over multiple SQL statements in a SQL string
+     @param {string} sql a string of SQL
+     @return {StatementIterator} the resulting statement iterator
+     */
+    Database.prototype["iterateStatements"] = function iterateStatements(sql) {
+        return new StatementIterator(sql, this);
+    };
+
+    /* Internal method to implement statement iteration */
+
+    Database.prototype["advanceIterator"] = function advanceIterator(iter) {
+        var stack = stackSave();
+        var pzTail = stackAlloc(4);
+        setValue(apiTemp, 0, "i32");
+        setValue(pzTail, 0, "i32");
+        var returnCode = sqlite3_prepare_v2(
+            this.db,
+            iter.nextSql,
+            -1,
+            apiTemp,
+            pzTail
+        );
+        var tail = UTF8ToString(getValue(pzTail, "i32"));
```

Comment:
This means we convert the string n times between utf8 and utf16, each time reallocating a new string. Can't we just allocate the string once and then work with pointers ?

Reply:
So I did this because in the next line I'm using the substring method - I wasn't sure that just doing the math with pointers would get the correct result in every instance.  In general I'm very shaky on how the runtime interacts with Javascript code.

Reply:
You shouldn't need to keep track of lastSql at all. You can use only nextSql and have it be a simple opaque pointer. You will need to create a new cwrap for  sqlite3_prepare_v2 that takes a number instead of a string.

Reply:
Hm, I'm struggling to understand this, sorry!  It seems like I need nextSql as a string, because I get a string in to start with.  Or can I convert it to a pointer for storage in the iterator?  However, I believe I still need a lastSql member for the client to access when there is an error.  The nextSql value contains the tail - the yet-to-be parsed and executed stuff, while the lastSql contains just the last executed substring.  And I would think it needs to be a string, or I need to have a method that converts it to a string for the clients.

Reply:
What you are currently doing is allocating memory, converting your string to utf8, copying it to the allocated memory, giving a pointer to the allocated memory to sqlite (all of this is handled by emscripten in cwrap), and then, when it returns, converting back the end of the string to utf16, and starting all over again. 

What we should actually be doing is allocating the string manually at the beginning, keep a simple pointer of the remaining text to be parsed, and updating it as we go. In the end, we manually free the entire string. 

We can add a method to convert the remaining text pointer to a string. 
In the current design, lastSql is not documented, so you could not rely on it anyway.

Reply:
When you say "at the beginning" and "in the end", what scope are you referring to?  The advanceIterator method, or the lifetime of the iterator object?

Yeah, I meant to document and test lastSql, I forgot about it.  What I need is not method to convert the *remaining* text pointer to a string, but a mechanism to retrieve the most recently attempted SQL.  I don't know how to do that if I don't store it.  I guess I could store *2* pointers and then a method to convert that stretch of bytes to a string.

To be more clear:
After the call to sqlite3_prepare_v2(), I have 1 pointer stored in pzTail.  Whether the call succeeds or fails, pzTail points to the portion of the SQL string *after* the statement that sqlite attempted to make a prepared statement from.  If we succeed, then we use that for the next iteration.  If we succeed, we can also retrieve the SQL from the prepared statement using sqlite3_sql().  However, if we fail, there is no statement, so no way to retrieve the attempted SQL except by using the original string and subtracting pzTail from it.  I guess to avoid the work you are describing, I need to keep a pointer to the last statement rather than the string, then I can use the span of bytes between the last pointer and pzTail to extract a string?  What function am I looking for that would let me turn a span of bytes between 2 pointers into a string (I can't just stick a null terminator in there without overwriting something)?

Reply:
I did some more reading on emscripten; do I want something like:
```
return HEAPU8.subarray(last_ptr, pzTail);
```

?

Reply:
> What I need is not method to convert the remaining text pointer to a string, but a mechanism to retrieve the most recently attempted SQL.

The most recently attempted SQL **is** the remaining text. 

> [...] After the call to sqlite3_prepare_v2(), I have 1 pointer stored in pzTail. Whether the call succeeds or fails, pzTail points to the portion of the SQL string after the statement that sqlite attempted to make a prepared statement from. If we succeed, then we use that for the next iteration. If we succeed, we can also retrieve the SQL from the prepared statement using sqlite3_sql(). However, if we fail, there is no statement, so no way to retrieve the attempted SQL except by using the original string and subtracting pzTail from it. 

This is not what the SQLite documentation says.
SQLite will try to parse what you give, and stop when it **either** successfully parses a statement **or** encounters an error. If the string you provided does not start with a valid sql statement, I think there is nothing you can do but just retrieve the full remaining text. SQLite does not make any guarantee to parse invalid statements correctly. Or at least that's what I understand from the documentation: https://sqlite.org/c3ref/prepare.html

>  I guess I could store 2 pointers and then a method to convert that stretch of bytes to a string.

You need two pointers: one to the full string to be able to free it once the iteration is over, and one to the currently remaining part of the string (pzTail). There is no notion of an end of an invalid sql statement.

Reply:
> SQLite will try to parse what you give, and stop when it **either** successfully parses a statement **or** encounters an error. If the string you provided does not start with a valid sql statement, I think there is nothing you can do but just retrieve the full remaining text. SQLite does not make any guarantee to parse invalid statements correctly. Or at least that's what I understand from the documentation: https://sqlite.org/c3ref/prepare.html

I see.  I read the docs slightly differently and tested with a certain case, and it behaved as I expected, so I didn't dig further.  Specifically, if you do something like this:
"create table test (x text); select foobar from test; select * from test"
and run it through my code, the second statement throws an exception when prepared because foobar is not a column in test.  When that happens, " select * from test" is what is in pzTail.  So it *looked* like it was doing what I wanted it to.  However, doing further testing, I see that pzTail can actually start earlier if sqlite detects an error earlier in the query.  OK.

Reply:
> You need two pointers: one to the full string to be able to free it once the iteration is over, ...

Wouldn't we then need users to call a free() method or something on the iterator when they are finished with it?  Otherwise I don't know how we'd detect that the users are done with it.

Reply:
We free the string and invalidate the iterator before returning `{done:true}` to the user.

Reply:
In that case, then, a user wouldn't be able to just do a couple of iterations manually (calling next()) and stop, though, right?  Is that what we want?  Also, what if there is an exception when we call handleError()?  Do we need to wrap that in a try/catch, or just check the status code manually?

Reply:
> Also, what if there is an exception when we call handleError()? Do we need to wrap that in a try/catch, or just check the status code manually?

Sorry, I meant to say, won't we need to free the memory in this case as well?

Reply:
I think we want to call handleError in a try block, with a catch and a finally